src/vm_cpu.rs

```diff
@@ -95,6 +94,62 @@ pub struct VmCpuTrapState {
     pub htval: u64,
     pub htinst: u64,
 }
+#[derive(Default)]
+#[repr(C)]
+struct VmPmuState {
+    // Stores the mapping for configured PMU counters.
+    configured_counter_map: [u64; drivers::pmu::MAX_COUNTERS],
+    // Stores a bit-map of indexes into PmuCounterInfo for configured counters
+    configured_counters: u64,
+    // Stores a bit-map of indexes into PmuCounterInfo for started counters
```

Comment:
Do you really need two separate data structure to track the configured counters ?
Counter space is not huge. The maximum number of hardware counters possible is 32 only.

Reply:
Right, but think we do need a bitmap of configured counters, and their respective mappings?

Reply:
Currently you have two bitmap

1. For configured counters
2. For started counters

Two array indexed by counter id.

configured_counter_map & counter_values can be combined two one as you need store the configured event and value for a counter. You can track the state in a structure as well but you do have to iterate all of them to find out which one is configured/started. It shouldn't a big cost given that counter space is not huge.

If you use the array/structure route, you can just track 1 state per counter as @abrestic-rivos pointed out.

unconfigured ->configured->started->stopped->unconfigured

Of course, bitmap makes it easier for iterating through the list. You don't need to track the started state though if you want to keep it simple. Keep only the configured bitmap. Don't care about the started bitmap at all. As long as the event is configured, you can try to stop it. It will just fail.